pkg/controller/installation/bpf.go

```diff
@@ -0,0 +1,170 @@
+// Copyright (c) 2019-2023 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package installation
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+
+	"github.com/go-logr/logr"
+	operator "github.com/tigera/operator/api/v1"
+	crdv1 "github.com/tigera/operator/pkg/apis/crd.projectcalico.org/v1"
+	"github.com/tigera/operator/pkg/common"
+	"github.com/tigera/operator/pkg/controller/migration/convert"
+	"github.com/tigera/operator/pkg/render"
+	appsv1 "k8s.io/api/apps/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func bpfUpgradeDaemonsetEnvVar(r *ReconcileInstallation, ctx context.Context, install *operator.Installation, ds *appsv1.DaemonSet, fc *crdv1.FelixConfiguration, reqLogger logr.Logger) error {
+
+	// Query calico-node DS: if FELIX_BPFENABLED env var set and FC bpfEnabled unset then patch FC and quit.
+	patchFelixConfig, err := processDaemonsetEnvVar(r, ctx, ds, fc, reqLogger)
+	if err != nil {
+		return err
+	}
+
+	// Attempt to patch Felix Config now.
+	return patchFelixConfigurationImpl(r, ctx, install, fc, reqLogger, patchFelixConfig)
+}
+
+func bpfUpgradeWithoutDisruption(r *ReconcileInstallation, ctx context.Context, install *operator.Installation, ds *appsv1.DaemonSet, fc *crdv1.FelixConfiguration, reqLogger logr.Logger) error {
```

Comment:
I think this function is also blurring the lines of responsibility more than I generally like. 

- The logic which checks the annotation needs to be performed everywhere that we might patch BPF enabled, so should probably be commonized.
- I also expected the logic that determines whether not we should patch FelixConfiguration at all (i.e., whether the rollout is complete and BPF is enabled) would determine whether or not we call this function at all rather than happening at the end of this function. This has a a real consequence both in terms of code readability, but also in logic (discussed in another comment).

Reply:
Thanks @caseydavenport - RE: annotation logic has been commonized in new private helper function `bpfValidateAnnotations`.  NB: have also refactored code as per second bullet point - thanks